### DISCUSSION TITLE
Mermaid in next-mdx-remote

### DISCUSSION THREAD

**Discussion by aakanshaverma11** [vercel/next.js#36369] [Help]
I am trying to build an app where we need to pull the remotely located md/mdx files, and render them on the app. I am able to build it with the next-mdx-remote but I am facing two challenges:

1. How to integrate [mdx-mermaid](https://sjwall.github.io/mdx-mermaid/docs/intro/) in the current setup ?
2. Looks like I need to maintain the IA in the package as I cannot read the window.location.path. So how can I avoid maintaining the IA(slug) pull the file based on the window path ?

Any help will be really appreciated. Thanks!

**Comment by icyJoseph** [comment]
Regarding passing `mdx-mermaid`, assumming you are using `next-mdx-remote` serialize:

```ts
import mermaidPlugin from 'mdx-mermaid' // this might bring commonJS vs ESM problems, I am not 100% sure

export async function getStaticProps() {
  // MDX text - can be from a local file, database, anywhere
  const source = 'Some **mdx** text, with a component <Test />'
  const mdxSource = await serialize(source, { mdxOptions: { remarkPlugins: [mermaidPlugin] } })
  return { props: { source: mdxSource } }
}
```

I think this should work for your first point.

Could you explain what's `IA`? and what are you trying to do in point 2? With other words or a list of steps?

**Reply by icyJoseph** [reply]
@aakanshaverma11 did you manage to make any progress?

**Reply by jesualdo13** [reply]
Hi @icyJoseph 

I'm trying to implement this solution as well, but it's not working. 

<img width="983" alt="image" src="https://user-images.githubusercontent.com/8900049/210875756-f9114080-a061-4fb3-b973-01db3f8b2333.png">

**Comment by mohsinsr** [comment]
I fixed this issue by creating a separate component for mermaid

1 - `import mermaid from "mermaid";
import { useEffect } from "react";

mermaid.initialize({
  startOnLoad: true,
});

type MermaidProps = {
  readonly chart: string;
};

const Mermaid = ({ chart }: MermaidProps): JSX.Element => {


  useEffect(() => {
    mermaid.contentLoaded();
  }, [])

  return <div className="mermaid">{chart}</div>;
};

export default Mermaid;
`




2 - Then import that component using next/dynamic 
`const Mermaid = dynamic(() => import("../Mermaid"), { ssr: false })`

3 - then you can pass this MDXRemote
  <MDXRemote {...content} components={{ Mermaid }} />
  
  
  
  .. following the above code resolved my issue.

**Comment by bluprince13** [comment]
Thanks @mohsinsr

I ended up using the solution from https://github.com/datopian/datahub/blob/8feb87739ddaeb07272666d4c11d50774a583e57/packages/core/src/ui/Mermaid/Mermaid.tsx#L12

Here is my commit that adds mermaid support to my blog which uses next-mdx-remote - https://github.com/bluprince13/bluprince13-next/commit/1cb54a17abce8dde6a51563a0af3085285a225c9

One thing I realised after some experimentation is that the order in which `mdx-mermaid` plugin is inserted into `remakrPlugins` matters. In my case, it had to come before the `codesandbox` plugin which I'm also using.
